oolab/src/main/java/agh/ics/oop/SimulationEngine.java

```diff
@@ -1,30 +1,28 @@
 package agh.ics.oop;
 
-import java.util.ArrayList;
-import java.util.List;
-
 class SimulationEngine implements IEngine{
 
     private final MoveDirection[] directions;
     private final IWorldMap map;
-    private Vector2d[] positions;
-    private List<Animal> animals = new ArrayList<>();
 
     public SimulationEngine(MoveDirection[] directions, IWorldMap map, Vector2d[] positions){
         this.directions = directions;
         this.map = map;
-        this.positions = positions;
-        for (Vector2d position : positions) {
-            Animal animal = new Animal(map, position);
+        for (Vector2d p : positions) {
+            Animal animal=new Animal(map, p);
+            animal.addObserver((IPositionChangeObserver) map);
```

Comment:
Jeśli każda mapa rozszerzałaby ten interfejs to nie musielibyśmy rzutować -> interfejsy też mogą rozszerzać inne interfejsy.